### PR TITLE
Rework iterator/iterable/array access resolving in `for` loops

### LINKED ISSUE #7661: `Iterator` implementation on an abstract is ignored if the abstract has `length` and `@:arrayAccess`
This issue is created because of ee4df650f06dde54ecbdc19a6308763fcc65cebd
Following code is expected to trace `0, 1, 2` but instead it traces `0, -1, -2` with the latest dev haxe:
```haxe
class Test {
  static function main() {
    for(i in new Iterato(0, 3)) {
      trace(i);
    }
  }
}

abstract Iterato({current:Int, last:Int}) {
  public var length(get,never):Int;
  inline function get_length() return this.last - this.current;
  
  @:arrayAccess 
  public function get(i:Int) return this.current - i;
  
  public inline function new(start:Int, end:Int) {
    this = {current:start, last:end};
  }
  
  public inline function hasNext()
  	return this.current < this.last;
  
  public inline function next()
  	return this.current++;
}
```
http://try-haxe.mrcdk.com/#78d4B

### LINKED ISSUE #8072: [JS] @:forward abstract _(Array<_>) will not use an inline iterator unless it defines length+@:arrayAccess
Might be half-related to https://github.com/HaxeFoundation/haxe/issues/7661

```haxe
class Test {
  static function main() {
    var one:One = [1, 2];
    for (v in one) trace(v);
    var two:Two = [1, 2];
    for (v in two) trace(v);
  }
}
@:forward abstract One(Array<Int>) from Array<Int> {
  
}
@:forward abstract Two(Array<Int>) from Array<Int> {
  public var length(get, never):Int;
  inline function get_length() return this.length;
  @:arrayAccess inline function get(i:Int) return this[i];
}
```
produces
```js
Test.main = function() {
	var v = HxOverrides.iter([1,2]);
	while(v.hasNext()) console.log("root/program/Test.hx:4:",v.next());
	var two = [1,2];
	var _g = 0;
	while(_g < two.length) console.log("root/program/Test.hx:6:",two[_g++]);
};
```
so the compiler knows that the object is iterable because it inherited the iterator from Array, but uses HxOverrides.iter unless the abstract itself defines length and an array getter.

### LINKED ISSUE #8343: Unable to implement iterator via `using` for types with array access
There is an issue with C# target: `class String implements ArrayAccess<cs.Char16> {` so the compiler picks length + ArrayAccess first, which makes `iterator` via using impossible for `String`.

_Originally posted by @RealyUniqueName in https://github.com/HaxeFoundation/haxe/pull/8331#issuecomment-496244211_ (also see replies)

### PR DESCRIPTION
**Pull request by RealyUniqueName**
Fixes #7661
Fixes #8343
Fixes #8072
Unblocks #8331

### CODE REVIEW

File: src/typing/forLoop.ml [hunk 1]
```diff
@@ -53,27 +53,113 @@ module IterationKind = struct
...
+			let last_resort () =
+				array_access_result := Some (of_texpr_by_array_access ctx e p);
+				mk (TConst TNull) t_dynamic p
+			in
+			let e1,pt = check_iterator ~resume ~last_resort ctx "iterator" e p in
+			match !array_access_result with
+			| None -> (IteratorIterator,e1,pt)
+			| Some result -> result
+		in
+		if e.epos.pfile = "src/Main.hx" then begin
```
Comment (Simn) [inline src/typing/forLoop.ml]: :D

Comment (RealyUniqueName) [reply src/typing/forLoop.ml]: Oh, sorry )

**Comment by Simn** [comment]
This is difficult to review with the big code move. Feel free to merge if you're sure there's no duplicate typing going on.

**Comment by RealyUniqueName** [comment]
`type_field` may be invoked twice for the same expression. Does it count for duplicate typing?
First time with `do_resume = true` and the second time with whatever `resume` value passed to `check_iterator` function.
First: https://github.com/HaxeFoundation/haxe/pull/8344/files#diff-794964db9b5615c66796d9a4a21ae740R83
Second: https://github.com/HaxeFoundation/haxe/pull/8344/files#diff-794964db9b5615c66796d9a4a21ae740R87
It is needed to try `IteratorArrayAccess` _after_ an attempt to resolve `iterator` field, but _before_ emitting an error about an absent `iterator` field.

**Comment by RealyUniqueName** [comment]
> `type_field` is safe to call multiple times, we do that in multiple places.
> As long as we don't use any function that goes from `expr` to `texpr` we're good

**Pull request merged**